client/client_test.go

```diff
@@ -0,0 +1,238 @@
+	}
+}
+
+// TestAddFile tests the `AddFile` function.  Checks that the index is properly
+// written by the server, and that errors are properly returned when the file is
+// not valid.
+func TestAddFile(t *testing.T) {
+	client, dir := startTestClient(t)
+	defer os.RemoveAll(dir)
+
+	content := "This is a random test string, it is quite long, or not really"
+	if err := ioutil.WriteFile(filepath.Join(dir, "testFile"), []byte(content), 0666); err != nil {
+		t.Fatalf("error when writing test file: %s", err)
+	}
+
+	if err := client.AddFile(filepath.Join(dir, "testFile")); err != nil {
+		t.Fatalf("error when adding the file: %s", err)
+	}
+
+	if err := client.AddFile(filepath.Join(dir, "nonExisting")); !os.IsNotExist(err) {
+		t.Fatalf("no error returned for non-existing file")
+	}
+
+	fileNotInDir, err := ioutil.TempFile("", "tmpFile")
+	if err != nil {
+		t.Fatalf("error when creating temporary test file: %s", err)
+	}
+	defer os.Remove(fileNotInDir.Name())
+
+	if err := client.AddFile(fileNotInDir.Name()); err.Error() != "target path not within base path" {
+		t.Fatalf("error not properly returned for file not in the client directory")
+	}
+}
+
+// TestRenameFile tests the `RenameFile` function.  Checks the indexes are
+// properly renamed and errors returned when necessary.
+func TestRenameFile(t *testing.T) {
+	client, dir := startTestClient(t)
+	defer os.RemoveAll(dir)
+
+	content := "a random content"
+	if err := ioutil.WriteFile(filepath.Join(dir, "testRenameFile"), []byte(content), 0666); err != nil {
+		t.Fatalf("error when writing test file: %s", err)
+	}
+
+	if err := client.AddFile(filepath.Join(dir, "testRenameFile")); err != nil {
+		t.Fatalf("error when adding the file: %s", err)
+	}
+
+	if err := client.RenameFile(filepath.Join(dir, "testRenameFile"), filepath.Join(dir, "testRename")); err != nil {
+		t.Fatalf("error when renaming file: %s", err)
+	}
+
+	// Doing the renaming second time should fail, as the orignal file no longer
+	// exists
+	if err := client.RenameFile(filepath.Join(dir, "testRenameFile"), filepath.Join(dir, "testRename")); !os.IsNotExist(err) {
+		t.Fatalf("error not properly returned when renaming non-existing file")
+	}
+}
+
+// TestDeleteFile tests the `DeleteFile` function.  Checks the indexes are
+// properly deleted and errors returned when necessary.
+func TestDeleteFile(t *testing.T) {
+	client, dir := startTestClient(t)
+	defer os.RemoveAll(dir)
+
+	content := "a random content"
+	if err := ioutil.WriteFile(filepath.Join(dir, "testDeleteFile"), []byte(content), 0666); err != nil {
+		t.Fatalf("error when writing test file: %s", err)
+	}
+
+	if err := client.AddFile(filepath.Join(dir, "testDeleteFile")); err != nil {
+		t.Fatalf("error when adding the file: %s", err)
+	}
+
+	if err := client.DeleteFile(filepath.Join(dir, "testDeleteFile")); err != nil {
+		t.Fatalf("error when deleting file: %s", err)
+	}
+
+	// Doing the deleting second time should fail, as the file no longer exists
```

Comment:
Actually it's probably better to have all RPCs be idempotent if possible.  That is, `Delete` should always succeed, even if the file doesn't exist.  Just in case the RPC response was dropped, the client can re-send and won't get confused by an error.  (Plus it wouldn't give any info away.)